lib/dialects/mysql/index.js

```diff
@@ -190,7 +190,10 @@ Object.assign(Client_MySQL.prototype, {
       .finally(() => {
         // NOT returning this promise because we want to release the connection
         // in a non-blocking fashion
-        acquiringConn.then((conn) => this.releaseConnection(conn));
+        acquiringConn
+          .then((conn) => this.releaseConnection(conn))
+          // error handled in another chain of promises
+          .catch(() => {});
```

Comment:
I think this might just be obscuring/hiding the underlying problem.

I agree that there is a race between 2 separate promise chains, and one of them is releasing the connection first.  However, it seems like the code should be restructured in a manner where the "release the connection" responsibility occurs _after_ the race has been resolved.  That is, instead of doing this:

```
RACE
  - pool.acquire().catch(()=>  releaseConnection())
  - delay().catch(()=> releaseConnection())
```

The code should be structured like this:

```
try {
  RACE
    - pool.acquire()
    - delay()
} catch() {
  releaseConnection()
}
```

(Sry -- I'm being sloppy about the Promise notation.  The main concept is that you shouldn't have 2 promise chains that each take responsibility for releasing the connection.  Instead, you should conclude the race, and then release the promise chain in 1 location)

Reply:
we cant releaseConnection  in catch/finally, if pool.acquire() in try block, because we doesn't have reference to it

Reply:
Maybe pool.acquire should have timeout

Reply:
Ok, we can pass connectTimeout while acquiring connection in mysql: https://github.com/mysqljs/mysql#connection-options

Reply:
Also pg does not give this error because acquiring connection not limited by time

Reply:
So, Knex is also providing a pooling layer on top of MySQL via the Tarn library.  One of the "timeouts" that we are observing is actually coming from Tarn.  In particular, take a look at line 351 below:

https://github.com/knex/knex/blob/77df705189c712105cfab293ae26fff41de8e6ae/lib/client.js#L348-L357

Typically, the Tarn connection pool would already be populated with MySQL connections.  Hence, the "timeout" doesn't necessarily mean that a connection w/ MySQL timed out.  It could also mean:

1. The Tarn pool was depleted, and no connections were returned to the pool in time
2. The Tarn pool was destroyed by mistake.  (I already checked for this, but this did not appear to be the case)
3. Tarn keeps trying to allocate a new resource for the pool, but it's repeatedly failing.  Eventually, the "acquisition" timeout expires.  (This could be the case if MySQL connections are repeatedly being destroyed before Tarn has an opportunity to hand them off)

Reply:
we can request tarn.js add timeout/cancellation of resource request

Reply:
k -- new observation.  The testcase fails differently if you change the pool size to 2:

https://github.com/knex/knex/blob/77df705189c712105cfab293ae26fff41de8e6ae/test/integration/builder/additional.js#L877

```
  1) Integration Tests
       mysql | mysql
         Additional
           .timeout(ms, {cancel: true}) should throw error if cancellation cannot acquire connection:

      AssertionError: expected { Object (timeout, name, ...) } to deeply equal { Object (timeout, name, ...) }
      + expected - actual

       {
      -  "message": "Defined query timeout of 1ms exceeded when running query."
      +  "message": "After query timeout of 1ms exceeded, cancelling of query failed."
         "name": "KnexTimeoutError"
         "timeout": 1
       }
      
      at /Users/brianlauber/Documents/src/knex/test/integration/builder/additional.js:912:71
  From previous event:
      at Context.<anonymous> (test/integration/builder/additional.js:911:15)
      at processImmediate (internal/timers.js:439:21)
```

It seems that the code is attempting to establish 2 separate connections for some reason.  So, there are 2 separate timeouts occurring:

1. The "1 ms" timeout as instructed by the testcase itself
2. The "100 ms" timeout for the pool resource acquisition

Furthermore, take a look at line 181 below:

https://github.com/knex/knex/blob/77df705189c712105cfab293ae26fff41de8e6ae/lib/dialects/mysql/index.js#L174-L196

For whatever reason, the cancelation operation is also introducing a 100ms delay.  So, I suspect the following is happening:

1. The initial connection times out after roughly 1ms
2. For whatever reason, the cancelation operation is mistakenly establishing a 2nd connection.
3. MEANWHILE, it is waiting 100ms before actually canceling the first connection
4. By then, the Tarn resource acquisition has already timed out for the 2nd connection

Reply:
... which makes sense.  It looks like the test case is intended to handle exactly this scenario.  The issue is that it's then failing to handle the Tarn acquisition error itself.

Reply:
.timeout(1) is for initial query
timeout(100) is for query which cancelling initial query

Reply:
cancellation operation establishing second connection because first connection is hanged, we cannot queue 2 parallel operations in one connection

Reply:
pool size should be = 1, because test reproducing case of timeout acquiring connection for cancel query

Reply:
I see exactly how/why the issue is occurring.  I'm just not 100% sure what to do about it at this point.  🤔 

Basically: the `cancelQuery(..)` method is written so that it tries to "give up" if it's unable to establish a secondary connection after 100ms.  So, the sequence of events is:

1. It starts trying to establish a secondary connection.
2. After 100ms, it finally gives up.
3. HOWEVER, it still invokes the `.finally(..)` block.
4. The `.finally(..)` block calls `acquiringConn.then(..)` again, which establishes a _second_ Promise chain.
5. The Tarn resource acquisition timeout is never handled in this second Promise chain.

Interestingly, the Tarn resource acquisition error _is_ being handled by the call to `Promise.race(..)`.  Unfortunately, after the race has already been concluded, the second promise chain is established, and the error is dropped on the floor.

So, I guess my main question: why is the `cancelQuery(..)` method enforcing a 100ms timeout in the first place?  Shouldn't it just rely upon the pool's settings instead?

cc: @maximelkin  + @kibertoad

Reply:
@bywo

Reply:
@bywo + @maximelkin : So yeah: my main question is "Do we actually need the 100ms timeout that is baked into the `cancelQuery(..)` method?".  Ex: why not just do this?

```
  async cancelQuery(connectionToKill) {
    const conn = await this.acquireConnection();
    try {
      await this.query(conn, {
        method: 'raw',
        sql: 'KILL QUERY ?',
        bindings: [connectionToKill.threadId],
        options: {},
      });
    } finally {
      await this.releaseConnection(conn);
    }
  },
```

This eliminates the race condition and ensures that the Tarn resource allocation error can be raised/caught.

Reply:
FYI: Just created an alternative merge request w/ this proposed solution:

https://github.com/knex/knex/pull/3666

All of the tests seem to be passing on my local machine, at least.  We'll see what Travis-CI has to say about it.